# 문제 :병원 개/폐업 분류 예측 경진대회
##  병원 재무 데이터와 AI로 개업|폐업 예측 분석 | 분류 | Accuracy


## 목차

1. 문제 정의
2. 분석 패키지
3. 데이터 읽기
4. 데이터 소개
5. 데이터 전처리
6. 데이터 분석(EDA)
7. 모델 선택과 훈련
8. 모델 평가
9. 결과

## 1. 문제 정의 

공공 교통 위치 빅데이터와 AI로 버스 운행 시간 분석

## 2. 분석 패키지

아래와 같은 패키지를 불러왔습니다.

In [192]:
import os
import numpy as np
import pandas as pd

import numpy as np
from tensorflow.keras.models import Sequential # 케라스의 Sequential()을 임포트
from tensorflow.keras.layers import Dense # 케라스의 Dense()를 임포트
from tensorflow.keras import optimizers # 케라스의 옵티마이저를 임포트


import matplotlib.pyplot as plt # 그래프 시각화
import seaborn as sns # 그래프 시각화
from sklearn.preprocessing import LabelEncoder # 라벨 인코더

## 3. 데이터 읽기

In [800]:
train = pd.read_csv('data\\train.csv')
test = pd.read_csv('data\\test.csv')
sub = pd.read_csv("data\\submission_sample.csv")

## 4. 데이터 소개

In [801]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 58 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   inst_id             301 non-null    int64  
 1   OC                  301 non-null    object 
 2   sido                301 non-null    object 
 3   sgg                 301 non-null    int64  
 4   openDate            301 non-null    int64  
 5   bedCount            296 non-null    float64
 6   instkind            300 non-null    object 
 7   revenue1            293 non-null    float64
 8   salescost1          293 non-null    float64
 9   sga1                293 non-null    float64
 10  salary1             293 non-null    float64
 11  noi1                293 non-null    float64
 12  noe1                293 non-null    float64
 13  interest1           293 non-null    float64
 14  ctax1               293 non-null    float64
 15  profit1             293 non-null    float64
 16  liquidAs

In [802]:
train_arr = np.array(train.columns)
for i in range(len(train_arr)):
    print("변수이름 : {}\n유일값 : {}\n유일값개수 : {}\n".format(train_arr[i],
                                                     train[train_arr[i]].unique(), len( train[train_arr[i]].unique())))

변수이름 : inst_id
유일값 : [  1   3   4   7   9  11  14  15  16  17  18  19  22  25  26  27  33  34
  35  36  37  38  39  42  44  45  47  49  50  51  52  53  55  56  57  58
  59  60  61  62  63  65  67  69  70  71  72  73  74  75  77  81  82  83
  84  87  89  91  92  93  94  95  96  97  98  99 100 101 104 105 106 107
 108 109 110 112 113 115 116 118 119 120 125 127 128 129 132 134 135 136
 137 138 139 140 141 143 144 145 147 148 149 150 152 153 154 155 156 159
 160 162 163 164 166 168 169 170 171 173 175 176 177 178 179 180 182 183
 187 189 190 191 192 194 195 196 197 200 202 203 204 205 208 209 211 214
 215 216 217 218 226 227 230 231 232 233 234 235 236 237 238 240 242 244
 245 247 249 250 251 252 253 255 256 257 259 261 262 263 264 265 266 267
 268 269 270 271 272 273 274 275 276 277 278 279 280 281 283 284 285 286
 287 288 289 290 291 295 296 297 298 300 302 303 304 305 306 308 309 311
 312 313 314 315 316 317 318 319 320 321 322 323 324 325 327 328 329 331
 332 333 335 338 339 340 342 3

In [803]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 127 entries, 0 to 126
Data columns (total 58 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   inst_id             127 non-null    int64  
 1   OC                  0 non-null      float64
 2   sido                127 non-null    object 
 3   sgg                 127 non-null    int64  
 4   openDate            126 non-null    float64
 5   bedCount            119 non-null    float64
 6   instkind            125 non-null    object 
 7   revenue1            125 non-null    float64
 8   salescost1          125 non-null    float64
 9   sga1                125 non-null    float64
 10  salary1             125 non-null    float64
 11  noi1                125 non-null    float64
 12  noe1                125 non-null    float64
 13  interest1           125 non-null    float64
 14  ctax1               125 non-null    float64
 15  profit1             125 non-null    float64
 16  liquidAs

In [804]:
test_arr = np.array(test.columns)
for i in range(len(test_arr)):
    print("변수이름 : {}\n유일값 : {}\n유일값개수 : {}\n".format(test_arr[i],
                                                     test[test_arr[i]].unique(), len( test[test_arr[i]].unique())))

변수이름 : inst_id
유일값 : [  2   5   6   8  10  12  13  20  21  23  24  28  29  30  31  32  40  41
  43  46  48  54  64  66  68  76  78  79  80  85  86  88  90 102 103 111
 114 117 121 122 123 124 126 130 131 133 142 146 151 157 158 161 165 167
 172 174 181 184 185 186 188 193 198 199 201 206 207 210 212 213 219 220
 221 222 223 224 225 228 229 239 241 243 246 254 258 260 282 292 293 294
 299 301 307 310 326 330 334 336 337 341 343 348 368 370 374 377 380 382
 387 388 389 394 395 396 397 398 401 403 404 410 413 421 424 425 429 430
 431]
유일값개수 : 127

변수이름 : OC
유일값 : [nan]
유일값개수 : 1

변수이름 : sido
유일값 : ['incheon' 'jeju' 'jeonnam' 'busan' 'jeonbuk' 'gyeonggi' 'gyeongbuk'
 'daejeon' 'seoul' 'daegu' 'choongbuk' 'ulsan' 'gyeongnam' 'choongnam'
 'sejong']
유일값개수 : 15

변수이름 : sgg
유일값 : [139 149 103  71  26 122  31 147  95 169  91  50  36  63  64  29 160 170
 142  40  87  34 152  83 162 127  46 118 157 130  92  67  23 155 166 144
 135  15 177   8  62 151  65 136  37  72  32 172  53  59 159  85  25 161

- inst_id : 그대로 사용
- OC : train의 값을 숫자형으로 변경, test에 있는 OC는 예측해야할 값이므로 제거
- sido : 머신러닝 알고리즘이 숫자를 다루므로 숫자형으로 변환, test와 train 의 유일값이 다르므로 유의해서 encoding 해주기
- sgg : 전체 훈련데이터가 300개임에 비해 유일값이 155개 이므로 이를 구분하면 모델이 과적합이 될 수 있기 때문에 제거 
- openDate : 병원 설립기간은 폐업과 관련이 있을 수 있으므로 새로운 파생변수 생성, test에 결측치가 한개 있는데 총 test 데이터의 중간값으로 처리
- bedCount : 병상의 수는 병원의 규모로 생각할 수 있으므로 연속형 데이터로 판단하고 변수로 사용, train과 test의 결측치를 중간값으로 처리 
- instkind : 문자를 숫자형으로 변경, test에 있는 결측치를 bedCount에 의거한 값으로 대체 시킨다
- revenue1 ~ surplus2 : 금액과 관련된 연속형 데이터이므로 train에서는 결측치가 포함된 행을 모두 제거한 데이터를 훈련시키고, test데이터에서는 해당 결측값을 중앙값으로 대체
- employee1,2 : 결측치가 없는 값을 중앙값으로 대체후 문자형태를 숫자형으로 변경
- ownerChange : 결측치를 대체할 값을 찾기 어렵기 때문에 empty로 대체

## 5. 데이터 전처리


In [805]:
# oc

oc = train['OC'].value_counts()


open      286
 close     15
Name: OC, dtype: int64

In [806]:
print('개업중인 병원 ": {} '.format(oc['open']))

개업중인 병원 ": 286 


In [807]:
print('폐업한 병원 ": {} '.format(oc[' close']))

폐업한 병원 ": 15 


### 2 - 1) test 데이터에도 적용

In [808]:
test.drop('OC', axis = 1, inplace = True ) # test 데이터에는 ['OC']가 필요 없음

### 2 - 2) open과 close에 번호 분류한 후 정수값으로 변환하기

In [809]:
train['OC'] = train['OC'].map( {'open': 1, ' close': 0} ).astype(int)
train['OC']

0      1
1      1
2      1
3      1
4      1
      ..
296    1
297    0
298    0
299    0
300    1
Name: OC, Length: 301, dtype: int32

### 2 - 3) ['sido']의 값을 수치형으로 범주화 시키기

In [812]:
  train['sido'].unique()

array([ 2,  9,  8, 10,  0, 13, 15, 12,  1, 16,  4,  3,  7,  5,  6, 14])

In [813]:
sido_encoder = LabelEncoder() # 인코더 생성

sido = list(train['sido'].values) + list(test['sido'].values) 
sido_set = set(sido)
print(len(sido_set))
sido_encoder.fit(list(sido_set)) # 인코딩

# 모든 학습, 시험 데이터의 정류장 정보 치환
train['sido'] = sido_encoder.transform(train['sido'])
test['sido'] = sido_encoder.transform(test['sido'])
test['sido'].unique

17


<bound method Series.unique of 0      10
1      11
2      13
3       0
4      12
       ..
122    13
123     8
124     8
125    11
126     1
Name: sido, Length: 127, dtype: int64>

### 2 - 4) opendate

**2018년 폐업을 2017년 폐업으로 간주한다고 했으니 현재 train 데이터에 나타난 close는 18년 12월 31일까지 영업을 했다고 간주 함**   
= opendate와 18-12-13을 기준으로 연산하여 영업일이라는 파생변수를 생성

In [755]:
train['DATE'] = 20181231  # 파생변수 train['DATE']
train['DATE'] = train['DATE'].astype(str)
train['DATE'] = train['DATE'].str[0:4] + "-" + train['DATE'].str[4:6] + "-" + train['DATE'].str[6:8]
train['DATE'] = train['DATE'].astype('datetime64[ns]')
train['DATE']

0     2018-12-31
1     2018-12-31
2     2018-12-31
3     2018-12-31
4     2018-12-31
         ...    
293   2018-12-31
294   2018-12-31
295   2018-12-31
296   2018-12-31
299   2018-12-31
Name: DATE, Length: 296, dtype: datetime64[ns]

In [756]:
# train['openDate']의 값을 날짜형 데이터로 바꾸기 (굉장히 오래걸리고 찾는데 애를 씀)
train['openDate'] = train['openDate'].astype(str)
train['opendate'] = train['openDate'].str[0:4] + "-" + train['openDate'].str[4:6] + "-" + train['openDate'].str[6:8]
train['opendate'] = train['opendate'].astype('datetime64[ns]')

KeyError: 'openDate'

In [ ]:
train.drop('openDate', axis = 1, inplace = True)

In [757]:
train

inst_id  OC  sido  sgg  bedCount  instkind      revenue1   salescost1  \
0          1   1     2   73     175.0         5  4.217530e+09          0.0   
1          3   1     9   32     410.0         3           NaN          NaN   
2          4   1     8   89     468.0         5  1.004522e+09  515483669.0   
3          7   1    10  141     353.0         3  7.250734e+10          0.0   
4          9   1     9   32     196.0         3  4.904354e+10          0.0   
..       ...  ..   ...  ...       ...       ...           ...          ...   
293      418   1     8  170     109.0         4  4.303270e+09  242387555.0   
294      419   1     9  174     367.0         5  1.207963e+10  530924107.0   
295      420   1     8  177       0.0         5  5.500472e+09          0.0   
296      422   1    13  178     214.0         5  7.614697e+09  415347464.0   
299      427   0     8  169     180.0         5  2.116892e+09  268174808.0   

             sga1       salary1  ...  NCLiabilities2     longLoan2  \
0    3.961135e+09  2.033835e+09  ...    5.361169e+08  3.900000e+08   
1             NaN           NaN  ...             NaN           NaN   
2    4.472197e+08  2.964023e+08  ...    0.000000e+00  0.000000e+00   
3    7.067740e+10  3.178605e+10  ...    2.073641e+10  1.510000e+10   
4    4.765605e+10  2.446078e+10  ...    2.136001e+10  1.410803e+10   
..            ...           ...  ...             ...           ...   
293  3.455912e+09  2.228914e+09  ...    1.969747e+09  1.925500e+09   
294  9.521440e+09  5.691328e+09  ...    6.090000e+09  5.890000e+09   
295  4.883855e+09  2.653680e+09  ...    3.233945e+09  1.800000e+09   
296  5.903119e+09  3.467861e+09  ...    6.805570e+09  5.930000e+09   
299  3.286245e+09  2.231944e+09  ...    0.000000e+00  0.000000e+00   

        netAsset2      surplus2  employee1  employee2  ownerChange       DATE  \
0    2.619290e+09  1.271224e+09       62.0       64.0         same 2018-12-31   
1             NaN           NaN      801.0      813.0         same 2018-12-31   
2    0.000000e+00  0.000000e+00      234.0        1.0         same 2018-12-31   
3    1.295427e+10  7.740829e+09      663.0      663.0         same 2018-12-31   
4    5.561941e+06  9.025550e+09      206.0      197.0         same 2018-12-31   
..            ...           ...        ...        ...          ...        ...   
293  2.637783e+09  0.000000e+00       75.0       70.0         same 2018-12-31   
294  9.065227e+09  0.000000e+00      180.0      180.0         same 2018-12-31   
295  4.337819e+09  0.000000e+00       62.0       66.0         same 2018-12-31   
296  5.379502e+09  8.116416e+08      193.0      141.0         same 2018-12-31   
299  0.000000e+00  0.000000e+00      100.0        NaN          NaN 2018-12-31   

      opendate  days  
0   2007-12-28  4021  
1   1997-04-01  7944  
2   2016-12-28   733  
3   2000-08-14  6713  
4   2005-09-01  4869  
..         ...   ...  
293 2012-03-12  2485  
294 2010-11-04  2979  
295 2011-05-23  2779  
296 2005-02-11  5071  
299 2007-01-01  4382  

[296 rows x 60 columns]

In [758]:
train['days'] = train['DATE'] - train['opendate']

In [759]:
train

inst_id  OC  sido  sgg  bedCount  instkind      revenue1   salescost1  \
0          1   1     2   73     175.0         5  4.217530e+09          0.0   
1          3   1     9   32     410.0         3           NaN          NaN   
2          4   1     8   89     468.0         5  1.004522e+09  515483669.0   
3          7   1    10  141     353.0         3  7.250734e+10          0.0   
4          9   1     9   32     196.0         3  4.904354e+10          0.0   
..       ...  ..   ...  ...       ...       ...           ...          ...   
293      418   1     8  170     109.0         4  4.303270e+09  242387555.0   
294      419   1     9  174     367.0         5  1.207963e+10  530924107.0   
295      420   1     8  177       0.0         5  5.500472e+09          0.0   
296      422   1    13  178     214.0         5  7.614697e+09  415347464.0   
299      427   0     8  169     180.0         5  2.116892e+09  268174808.0   

             sga1       salary1  ...  NCLiabilities2     longLoan2  \
0    3.961135e+09  2.033835e+09  ...    5.361169e+08  3.900000e+08   
1             NaN           NaN  ...             NaN           NaN   
2    4.472197e+08  2.964023e+08  ...    0.000000e+00  0.000000e+00   
3    7.067740e+10  3.178605e+10  ...    2.073641e+10  1.510000e+10   
4    4.765605e+10  2.446078e+10  ...    2.136001e+10  1.410803e+10   
..            ...           ...  ...             ...           ...   
293  3.455912e+09  2.228914e+09  ...    1.969747e+09  1.925500e+09   
294  9.521440e+09  5.691328e+09  ...    6.090000e+09  5.890000e+09   
295  4.883855e+09  2.653680e+09  ...    3.233945e+09  1.800000e+09   
296  5.903119e+09  3.467861e+09  ...    6.805570e+09  5.930000e+09   
299  3.286245e+09  2.231944e+09  ...    0.000000e+00  0.000000e+00   

        netAsset2      surplus2  employee1  employee2  ownerChange       DATE  \
0    2.619290e+09  1.271224e+09       62.0       64.0         same 2018-12-31   
1             NaN           NaN      801.0      813.0         same 2018-12-31   
2    0.000000e+00  0.000000e+00      234.0        1.0         same 2018-12-31   
3    1.295427e+10  7.740829e+09      663.0      663.0         same 2018-12-31   
4    5.561941e+06  9.025550e+09      206.0      197.0         same 2018-12-31   
..            ...           ...        ...        ...          ...        ...   
293  2.637783e+09  0.000000e+00       75.0       70.0         same 2018-12-31   
294  9.065227e+09  0.000000e+00      180.0      180.0         same 2018-12-31   
295  4.337819e+09  0.000000e+00       62.0       66.0         same 2018-12-31   
296  5.379502e+09  8.116416e+08      193.0      141.0         same 2018-12-31   
299  0.000000e+00  0.000000e+00      100.0        NaN          NaN 2018-12-31   

      opendate      days  
0   2007-12-28 4021 days  
1   1997-04-01 7944 days  
2   2016-12-28  733 days  
3   2000-08-14 6713 days  
4   2005-09-01 4869 days  
..         ...       ...  
293 2012-03-12 2485 days  
294 2010-11-04 2979 days  
295 2011-05-23 2779 days  
296 2005-02-11 5071 days  
299 2007-01-01 4382 days  

[296 rows x 60 columns]

train['days']의 값을 정수형으로 바꾸기

In [760]:
train['days']

0     4021 days
1     7944 days
2      733 days
3     6713 days
4     4869 days
         ...   
293   2485 days
294   2979 days
295   2779 days
296   5071 days
299   4382 days
Name: days, Length: 296, dtype: timedelta64[ns]

In [761]:
train['days'] = train['days'].dt.days

In [762]:
train

inst_id  OC  sido  sgg  bedCount  instkind      revenue1   salescost1  \
0          1   1     2   73     175.0         5  4.217530e+09          0.0   
1          3   1     9   32     410.0         3           NaN          NaN   
2          4   1     8   89     468.0         5  1.004522e+09  515483669.0   
3          7   1    10  141     353.0         3  7.250734e+10          0.0   
4          9   1     9   32     196.0         3  4.904354e+10          0.0   
..       ...  ..   ...  ...       ...       ...           ...          ...   
293      418   1     8  170     109.0         4  4.303270e+09  242387555.0   
294      419   1     9  174     367.0         5  1.207963e+10  530924107.0   
295      420   1     8  177       0.0         5  5.500472e+09          0.0   
296      422   1    13  178     214.0         5  7.614697e+09  415347464.0   
299      427   0     8  169     180.0         5  2.116892e+09  268174808.0   

             sga1       salary1  ...  NCLiabilities2     longLoan2  \
0    3.961135e+09  2.033835e+09  ...    5.361169e+08  3.900000e+08   
1             NaN           NaN  ...             NaN           NaN   
2    4.472197e+08  2.964023e+08  ...    0.000000e+00  0.000000e+00   
3    7.067740e+10  3.178605e+10  ...    2.073641e+10  1.510000e+10   
4    4.765605e+10  2.446078e+10  ...    2.136001e+10  1.410803e+10   
..            ...           ...  ...             ...           ...   
293  3.455912e+09  2.228914e+09  ...    1.969747e+09  1.925500e+09   
294  9.521440e+09  5.691328e+09  ...    6.090000e+09  5.890000e+09   
295  4.883855e+09  2.653680e+09  ...    3.233945e+09  1.800000e+09   
296  5.903119e+09  3.467861e+09  ...    6.805570e+09  5.930000e+09   
299  3.286245e+09  2.231944e+09  ...    0.000000e+00  0.000000e+00   

        netAsset2      surplus2  employee1  employee2  ownerChange       DATE  \
0    2.619290e+09  1.271224e+09       62.0       64.0         same 2018-12-31   
1             NaN           NaN      801.0      813.0         same 2018-12-31   
2    0.000000e+00  0.000000e+00      234.0        1.0         same 2018-12-31   
3    1.295427e+10  7.740829e+09      663.0      663.0         same 2018-12-31   
4    5.561941e+06  9.025550e+09      206.0      197.0         same 2018-12-31   
..            ...           ...        ...        ...          ...        ...   
293  2.637783e+09  0.000000e+00       75.0       70.0         same 2018-12-31   
294  9.065227e+09  0.000000e+00      180.0      180.0         same 2018-12-31   
295  4.337819e+09  0.000000e+00       62.0       66.0         same 2018-12-31   
296  5.379502e+09  8.116416e+08      193.0      141.0         same 2018-12-31   
299  0.000000e+00  0.000000e+00      100.0        NaN          NaN 2018-12-31   

      opendate  days  
0   2007-12-28  4021  
1   1997-04-01  7944  
2   2016-12-28   733  
3   2000-08-14  6713  
4   2005-09-01  4869  
..         ...   ...  
293 2012-03-12  2485  
294 2010-11-04  2979  
295 2011-05-23  2779  
296 2005-02-11  5071  
299 2007-01-01  4382  

[296 rows x 60 columns]

### 2 - 5) test 데이터도 같은 작업

In [763]:
test['DATE'] = 20181231  # 파생변수 train['DATE']
test['DATE'] = test['DATE'].astype(str)
test['DATE'] = test['DATE'].str[0:4] + "-" + test['DATE'].str[4:6] + "-" + test['DATE'].str[6:8]
test['DATE'] = test['DATE'].astype('datetime64[ns]')
test['DATE']

0     2018-12-31
1     2018-12-31
2     2018-12-31
3     2018-12-31
4     2018-12-31
         ...    
122   2018-12-31
123   2018-12-31
124   2018-12-31
125   2018-12-31
126   2018-12-31
Name: DATE, Length: 127, dtype: datetime64[ns]

### 

In [764]:
test['openDate'][120] # 결측치를 특정값으로 대체

'20011012.0'

In [765]:
test['openDate'][120] = 20011012.0
test.tail(20)

C:\Python\anaconda3\envs\mldl\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


inst_id  sido  sgg    openDate  bedCount  instkind      revenue1  \
107      382    16  123  19970422.0     103.0         4  8.159246e+10   
108      387     7  132  20060401.0      74.0         5  2.755846e+09   
109      388     7  132  20000801.0     102.0         4  9.737271e+09   
110      389     8  134  20120503.0      53.0         4  6.296805e+09   
111      394    10  140  19990610.0      20.0         0  1.093225e+09   
112      395    12  143  20161229.0     156.0         5  1.061103e+07   
113      396    12  143  20161229.0     156.0         5  1.061103e+07   
114      397    12  147  20130722.0     194.0         5  5.479248e+09   
115      398    11  149  20070912.0      13.0         5  7.492834e+09   
116      401     9  155  20080602.0     170.0         4  7.897861e+09   
117      403     2  159  20131226.0     148.0         5  4.094378e+09   
118      404     2  159  20110926.0     149.0         5  1.713748e+09   
119      410     1  165  19930614.0     136.0         5  3.094658e+09   
120      413     8  168  2.0011e+07      49.0         2  5.583625e+08   
121      421     8  177  20110415.0     167.0         5  4.471030e+08   
122      424    13   33  20011012.0       0.0         6  2.233031e+10   
123      425     8   74  20010530.0     156.0         7  1.833906e+10   
124      429     8  116  20010101.0     150.0         4  1.424266e+10   
125      430    11   76  20010201.0     156.0         2  4.892710e+10   
126      431     1  163  19961216.0      96.0         4  1.092444e+10   

       salescost1          sga1       salary1  ...  NCLiabilities2  \
107  1.495838e+10  5.838249e+10  4.118066e+10  ...    2.357855e+10   
108  0.000000e+00  2.691651e+09  1.561923e+09  ...    2.030000e+09   
109  0.000000e+00  0.000000e+00  0.000000e+00  ...    3.463048e+09   
110  1.192329e+09  4.880758e+09  3.477672e+09  ...    0.000000e+00   
111  0.000000e+00  1.064146e+09  6.455580e+08  ...    4.100000e+07   
112  0.000000e+00  8.800000e+03  0.000000e+00  ...    0.000000e+00   
113  0.000000e+00  8.800000e+03  0.000000e+00  ...    0.000000e+00   
114  2.959746e+08  5.173015e+09  3.263054e+09  ...    2.092021e+09   
115  0.000000e+00  6.547070e+09  2.527670e+09  ...    6.457484e+09   
116  8.668840e+08  6.393699e+09  3.553658e+09  ...    3.150323e+09   
117  0.000000e+00  4.062433e+09  2.025038e+09  ...    0.000000e+00   
118  1.540489e+08  1.544841e+09  9.090099e+08  ...    0.000000e+00   
119  2.227340e+08  2.657350e+09  1.980105e+09  ...    4.033884e+09   
120  7.443415e+07  5.482900e+08  2.826852e+08  ...    0.000000e+00   
121  1.423339e+08  2.581514e+08  1.191270e+08  ...    9.731573e+08   
122  8.484657e+08  1.849255e+10  1.232241e+10  ...    2.750371e+09   
123  0.000000e+00  1.760117e+10  6.824241e+09  ...    1.799886e+09   
124  1.726174e+09  1.148722e+10  6.279155e+09  ...    0.000000e+00   
125  4.157148e+10  4.721485e+09  1.514547e+09  ...    2.871805e+10   
126  1.168731e+09  8.709839e+09  3.784543e+09  ...    0.000000e+00   

        longLoan2     netAsset2      surplus2  employee1  employee2  \
107  1.479106e+10  4.077732e+10  2.377131e+10        604        604   
108  2.030000e+09  2.030865e+09  0.000000e+00         55         56   
109  2.422436e+09 -9.232669e+08  0.000000e+00         33         20   
110  0.000000e+00  0.000000e+00  0.000000e+00         63         63   
111  4.100000e+07  1.052438e+09  0.000000e+00         16         16   
112  0.000000e+00  0.000000e+00  0.000000e+00         36        NaN   
113  0.000000e+00  0.000000e+00  0.000000e+00         36        NaN   
114  2.092021e+09  5.475197e+09  0.000000e+00        106        103   
115  5.786636e+09  3.217245e+09  0.000000e+00         66         68   
116  3.150323e+09  5.881965e+09  1.299147e+09        104         98   
117  0.000000e+00  0.000000e+00  0.000000e+00         60         57   
118  0.000000e+00  2.109904e+09  0.000000e+00         19         19   
119  3.005151e+09  2.661298e+09  0.000000e+00         49         40   
120  0

In [766]:
# train['openDate']의 값을 날짜형 데이터로 바꾸기 (굉장히 오래걸리고 찾는데 애를 씀)
test['openDate'] = test['openDate'].astype(str)
test['opendate'] = test['openDate'].str[0:4] + "-" + test['openDate'].str[4:6] + "-" + test['openDate'].str[6:8]
test['opendate'] = test['opendate'].astype('datetime64[ns]')

In [767]:
test['opendate']

0     1998-11-25
1     2016-03-09
2     1989-04-27
3     2010-02-26
4     2004-06-04
         ...    
122   2001-10-12
123   2001-05-30
124   2001-01-01
125   2001-02-01
126   1996-12-16
Name: opendate, Length: 127, dtype: datetime64[ns]

In [768]:
test['days'] = test['DATE'] - test['opendate']

In [769]:
test.tail(20)

inst_id  sido  sgg    openDate  bedCount  instkind      revenue1  \
107      382    16  123  19970422.0     103.0         4  8.159246e+10   
108      387     7  132  20060401.0      74.0         5  2.755846e+09   
109      388     7  132  20000801.0     102.0         4  9.737271e+09   
110      389     8  134  20120503.0      53.0         4  6.296805e+09   
111      394    10  140  19990610.0      20.0         0  1.093225e+09   
112      395    12  143  20161229.0     156.0         5  1.061103e+07   
113      396    12  143  20161229.0     156.0         5  1.061103e+07   
114      397    12  147  20130722.0     194.0         5  5.479248e+09   
115      398    11  149  20070912.0      13.0         5  7.492834e+09   
116      401     9  155  20080602.0     170.0         4  7.897861e+09   
117      403     2  159  20131226.0     148.0         5  4.094378e+09   
118      404     2  159  20110926.0     149.0         5  1.713748e+09   
119      410     1  165  19930614.0     136.0         5  3.094658e+09   
120      413     8  168  20011012.0      49.0         2  5.583625e+08   
121      421     8  177  20110415.0     167.0         5  4.471030e+08   
122      424    13   33  20011012.0       0.0         6  2.233031e+10   
123      425     8   74  20010530.0     156.0         7  1.833906e+10   
124      429     8  116  20010101.0     150.0         4  1.424266e+10   
125      430    11   76  20010201.0     156.0         2  4.892710e+10   
126      431     1  163  19961216.0      96.0         4  1.092444e+10   

       salescost1          sga1       salary1  ...  NCLiabilities2  \
107  1.495838e+10  5.838249e+10  4.118066e+10  ...    2.357855e+10   
108  0.000000e+00  2.691651e+09  1.561923e+09  ...    2.030000e+09   
109  0.000000e+00  0.000000e+00  0.000000e+00  ...    3.463048e+09   
110  1.192329e+09  4.880758e+09  3.477672e+09  ...    0.000000e+00   
111  0.000000e+00  1.064146e+09  6.455580e+08  ...    4.100000e+07   
112  0.000000e+00  8.800000e+03  0.000000e+00  ...    0.000000e+00   
113  0.000000e+00  8.800000e+03  0.000000e+00  ...    0.000000e+00   
114  2.959746e+08  5.173015e+09  3.263054e+09  ...    2.092021e+09   
115  0.000000e+00  6.547070e+09  2.527670e+09  ...    6.457484e+09   
116  8.668840e+08  6.393699e+09  3.553658e+09  ...    3.150323e+09   
117  0.000000e+00  4.062433e+09  2.025038e+09  ...    0.000000e+00   
118  1.540489e+08  1.544841e+09  9.090099e+08  ...    0.000000e+00   
119  2.227340e+08  2.657350e+09  1.980105e+09  ...    4.033884e+09   
120  7.443415e+07  5.482900e+08  2.826852e+08  ...    0.000000e+00   
121  1.423339e+08  2.581514e+08  1.191270e+08  ...    9.731573e+08   
122  8.484657e+08  1.849255e+10  1.232241e+10  ...    2.750371e+09   
123  0.000000e+00  1.760117e+10  6.824241e+09  ...    1.799886e+09   
124  1.726174e+09  1.148722e+10  6.279155e+09  ...    0.000000e+00   
125  4.157148e+10  4.721485e+09  1.514547e+09  ...    2.871805e+10   
126  1.168731e+09  8.709839e+09  3.784543e+09  ...    0.000000e+00   

        longLoan2     netAsset2      surplus2  employee1  employee2  \
107  1.479106e+10  4.077732e+10  2.377131e+10        604        604   
108  2.030000e+09  2.030865e+09  0.000000e+00         55         56   
109  2.422436e+09 -9.232669e+08  0.000000e+00         33         20   
110  0.000000e+00  0.000000e+00  0.000000e+00         63         63   
111  4.100000e+07  1.052438e+09  0.000000e+00         16         16   
112  0.000000e+00  0.000000e+00  0.000000e+00         36        NaN   
113  0.000000e+00  0.000000e+00  0.000000e+00         36        NaN   
114  2.092021e+09  5.475197e+09  0.000000e+00        106        103   
115  5.786636e+09  3.217245e+09  0.000000e+00         66         68   
116  3.150323e+09  5.881965e+09  1.299147e+09        104         98   
117  0.000000e+00  0.000000e+00  0.000000e+00         60         57   
118  0.000000e+00  2.109904e+09  0.000000e+00         19         19   
119  3.005151e+09  2.661298e+09  0.000000e+00         49         40   
120  0

In [770]:
test['days']

0      7341 days
1      1027 days
2     10840 days
3      3230 days
4      5323 days
         ...    
122    6289 days
123    6424 days
124    6573 days
125    6542 days
126    8050 days
Name: days, Length: 127, dtype: timedelta64[ns]

In [771]:
test['days'] = test['days'].dt.days

In [772]:
test

inst_id  sido  sgg    openDate  bedCount  instkind      revenue1  \
0          2    10  139  19981125.0     300.0         3  6.682486e+10   
1          5    11  149  20160309.0      44.0         4  3.495758e+10   
2          6    13  103  19890427.0     276.0         3  2.326031e+10   
3          8     0   71  20100226.0     363.0         3  0.000000e+00   
4         10    12   26  20040604.0     213.0         3  5.037025e+10   
..       ...   ...  ...         ...       ...       ...           ...   
122      424    13   33  20011012.0       0.0         6  2.233031e+10   
123      425     8   74  20010530.0     156.0         7  1.833906e+10   
124      429     8  116  20010101.0     150.0         4  1.424266e+10   
125      430    11   76  20010201.0     156.0         2  4.892710e+10   
126      431     1  163  19961216.0      96.0         4  1.092444e+10   

       salescost1          sga1       salary1  ...  NCLiabilities2  \
0    0.000000e+00  6.565709e+10  3.223695e+10  ...    4.720000e+09   
1    0.000000e+00  3.259270e+10  1.672254e+10  ...    2.521009e+10   
2    2.542571e+09  2.308749e+10  1.168734e+10  ...    0.000000e+00   
3    0.000000e+00  0.000000e+00  0.000000e+00  ...    2.558931e+09   
4    0.000000e+00  4.855803e+10  2.296346e+10  ...    3.642417e+10   
..            ...           ...           ...  ...             ...   
122  8.484657e+08  1.849255e+10  1.232241e+10  ...    2.750371e+09   
123  0.000000e+00  1.760117e+10  6.824241e+09  ...    1.799886e+09   
124  1.726174e+09  1.148722e+10  6.279155e+09  ...    0.000000e+00   
125  4.157148e+10  4.721485e+09  1.514547e+09  ...    2.871805e+10   
126  1.168731e+09  8.709839e+09  3.784543e+09  ...    0.000000e+00   

        longLoan2     netAsset2      surplus2  employee1  employee2  \
0    4.690000e+09  1.608540e+10  8.944587e+09        693        693   
1    1.830000e+10  3.789135e+09  0.000000e+00        379        371   
2    0.000000e+00  0.000000e+00  1.638540e+10        NaN        NaN   
3    0.000000e+00  3.914284e+10  0.000000e+00        760        760   
4    2.134712e+10  2.574488e+10  1.507269e+10        437        385   
..            ...           ...           ...        ...        ...   
122  0.000000e+00  1.384958e+10  0.000000e+00        560        NaN   
123  1.740000e+09  8.574502e+09  7.136681e+09        132        137   
124  0.000000e+00  0.000000e+00  0.000000e+00        NaN        NaN   
125  2.563120e+10 -2.050629e+08  0.000000e+00        363        343   
126  0.000000e+00  0.000000e+00  0.000000e+00        NaN        NaN   

     ownerChange       DATE   opendate   days  
0           same 2018-12-31 1998-11-25   7341  
1           same 2018-12-31 2016-03-09   1027  
2            NaN 2018-12-31 1989-04-27  10840  
3           same 2018-12-31 2010-02-26   3230  
4           same 2018-12-31 2004-06-04   5323  
..           ...        ...        ...    ...  
122          NaN 2018-12-31 2001-10-12   6289  
123         same 2018-12-31 2001-05-30   6424  
124          NaN 2018-12-31 2001-01-01   6573  
125         same 2018-12-31 2001-02-01   6542  
126          NaN 2018-12-31 1996-12-16   8050  

[127 rows x 60 columns]

opendate가 결측치 였던 행의 days 값을 test 데이터의 평균값으로 대체

In [773]:
test['days'].mean()

4928.314960629921

In [774]:
test['days'][120] = 4928.314960629921

C:\Python\anaconda3\envs\mldl\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [775]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296 entries, 0 to 299
Data columns (total 60 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   inst_id             296 non-null    int64         
 1   OC                  296 non-null    int32         
 2   sido                296 non-null    int64         
 3   sgg                 296 non-null    int64         
 4   bedCount            296 non-null    float64       
 5   instkind            296 non-null    int32         
 6   revenue1            288 non-null    float64       
 7   salescost1          288 non-null    float64       
 8   sga1                288 non-null    float64       
 9   salary1             288 non-null    float64       
 10  noi1                288 non-null    float64       
 11  noe1                288 non-null    float64       
 12  interest1           288 non-null    float64       
 13  ctax1               288 non-null    float64       

In [776]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 127 entries, 0 to 126
Data columns (total 60 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   inst_id             127 non-null    int64         
 1   sido                127 non-null    int64         
 2   sgg                 127 non-null    int64         
 3   openDate            127 non-null    object        
 4   bedCount            127 non-null    float64       
 5   instkind            127 non-null    int32         
 6   revenue1            125 non-null    float64       
 7   salescost1          125 non-null    float64       
 8   sga1                125 non-null    float64       
 9   salary1             125 non-null    float64       
 10  noi1                125 non-null    float64       
 11  noe1                125 non-null    float64       
 12  interest1           125 non-null    float64       
 13  ctax1               125 non-null    float64       

In [777]:
# bedCount 결측치 처리
train = train[train['bedCount'].notnull()]
test['bedCount'] = test['bedCount'].fillna(test[test['bedCount'].notnull()]['bedCount'].median())

In [778]:
# instkind 결측치 처리

train = train[train['instkind'].notnull()]
test['instkind'] = test['instkind'].fillna('empty')
instkind_encoder = LabelEncoder() # 인코더 생성

_instkind = list(train['instkind'].values) + list(test['instkind'].values) 
instkind_set = set(_instkind)
instkind_encoder.fit(list(instkind_set)) # 인코딩

# train과 test에 숫자형 치환
train['instkind'] = instkind_encoder.transform(train['instkind'])
test['instkind'] = instkind_encoder.transform(test['instkind'])

In [779]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 296 entries, 0 to 299
Data columns (total 60 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   inst_id             296 non-null    int64         
 1   OC                  296 non-null    int32         
 2   sido                296 non-null    int64         
 3   sgg                 296 non-null    int64         
 4   bedCount            296 non-null    float64       
 5   instkind            296 non-null    int64         
 6   revenue1            288 non-null    float64       
 7   salescost1          288 non-null    float64       
 8   sga1                288 non-null    float64       
 9   salary1             288 non-null    float64       
 10  noi1                288 non-null    float64       
 11  noe1                288 non-null    float64       
 12  interest1           288 non-null    float64       
 13  ctax1               288 non-null    float64       

In [734]:
# revenue1 ~ surplus2 결측치 처리

In [780]:
# train 

r1s2_drop = train.fillna(0)

In [781]:
missing_fill_val = {'revenue1': r1s2_drop.revenue1.mean(),
                    'salescost1': r1s2_drop.salescost1.mean(),
                    'sga1': r1s2_drop.sga1.mean(),
                    'salary1': r1s2_drop.salary1.mean(),
                    'noi1': r1s2_drop.noi1.mean(),
                    'noe1': r1s2_drop.noe1.mean(),
                    'interest1': r1s2_drop.interest1.mean(),
                    'ctax1': r1s2_drop.ctax1.mean(),
                    'profit1': r1s2_drop.profit1.mean(),
                    'liquidAsset1': r1s2_drop.liquidAsset1.mean(),
                    'quickAsset1': r1s2_drop.quickAsset1.mean(),
                    'receivableS1': r1s2_drop.receivableS1.mean(),
                    'inventoryAsset1': r1s2_drop.inventoryAsset1.mean(),
                    'nonCAsset1': r1s2_drop.nonCAsset1.mean(),
                    'tanAsset1': r1s2_drop.tanAsset1.mean(),
                    'OnonCAsset1': r1s2_drop.OnonCAsset1.mean(),
                    'receivableL1': r1s2_drop.receivableL1.mean(),
                    'debt1': r1s2_drop.debt1.mean(),
                    'liquidLiabilities1': r1s2_drop.liquidLiabilities1.mean(),
                    'shortLoan1': r1s2_drop.shortLoan1.mean(),
                    'NCLiabilities1': r1s2_drop.NCLiabilities1.mean(),
                    'longLoan1': r1s2_drop.longLoan1.mean(),
                    'netAsset1': r1s2_drop.netAsset1.mean(),
                    'surplus1': r1s2_drop.surplus1.mean(),
                    'revenue2': r1s2_drop.revenue2.mean(),
                    'salescost2': r1s2_drop.salescost2.mean(),
                    'sga2': r1s2_drop.sga2.mean(),
                    'salary2': r1s2_drop.salary2.mean(),
                    'noi2': r1s2_drop.noi2.mean(),
                    'noe2': r1s2_drop.noe2.mean(),
                    'interest2': r1s2_drop.interest2.mean(),
                    'ctax2': r1s2_drop.ctax2.mean(),
                    'profit2': r1s2_drop.profit2.mean(),
                    'liquidAsset2': r1s2_drop.liquidAsset2.mean(),
                    'quickAsset2': r1s2_drop.quickAsset2.mean(),
                    'receivableS2': r1s2_drop.receivableS2.mean(),
                    'inventoryAsset2': r1s2_drop.inventoryAsset2.mean(),
                    'nonCAsset2': r1s2_drop.nonCAsset2.mean(),
                    'tanAsset2': r1s2_drop.tanAsset2.mean(),
                    'OnonCAsset2': r1s2_drop.OnonCAsset2.mean(),
                    'receivableL2': r1s2_drop.receivableL2.mean(),
                    'debt2': r1s2_drop.debt2.mean(),
                    'liquidLiabilities2': r1s2_drop.liquidLiabilities2.mean(),
                    'shortLoan2': r1s2_drop.shortLoan2.mean(),
                    'NCLiabilities2': r1s2_drop.NCLiabilities2.mean(),
                    'longLoan2': r1s2_drop.longLoan2.mean(),
                    'netAsset2': r1s2_drop.netAsset2.mean(),
                    'surplus2': r1s2_drop.surplus2.mean()}

In [782]:
train2 = r1s2_drop.replace(0, missing_fill_val)

In [783]:
# test

r1s2_drop = test.fillna(0)

In [784]:
missing_fill_val = {'revenue1': r1s2_drop.revenue1.mean(),
                    'salescost1': r1s2_drop.salescost1.mean(),
                    'sga1': r1s2_drop.sga1.mean(),
                    'salary1': r1s2_drop.salary1.mean(),
                    'noi1': r1s2_drop.noi1.mean(),
                    'noe1': r1s2_drop.noe1.mean(),
                    'interest1': r1s2_drop.interest1.mean(),
                    'ctax1': r1s2_drop.ctax1.mean(),
                    'profit1': r1s2_drop.profit1.mean(),
                    'liquidAsset1': r1s2_drop.liquidAsset1.mean(),
                    'quickAsset1': r1s2_drop.quickAsset1.mean(),
                    'receivableS1': r1s2_drop.receivableS1.mean(),
                    'inventoryAsset1': r1s2_drop.inventoryAsset1.mean(),
                    'nonCAsset1': r1s2_drop.nonCAsset1.mean(),
                    'tanAsset1': r1s2_drop.tanAsset1.mean(),
                    'OnonCAsset1': r1s2_drop.OnonCAsset1.mean(),
                    'receivableL1': r1s2_drop.receivableL1.mean(),
                    'debt1': r1s2_drop.debt1.mean(),
                    'liquidLiabilities1': r1s2_drop.liquidLiabilities1.mean(),
                    'shortLoan1': r1s2_drop.shortLoan1.mean(),
                    'NCLiabilities1': r1s2_drop.NCLiabilities1.mean(),
                    'longLoan1': r1s2_drop.longLoan1.mean(),
                    'netAsset1': r1s2_drop.netAsset1.mean(),
                    'surplus1': r1s2_drop.surplus1.mean(),
                    'revenue2': r1s2_drop.revenue2.mean(),
                    'salescost2': r1s2_drop.salescost2.mean(),
                    'sga2': r1s2_drop.sga2.mean(),
                    'salary2': r1s2_drop.salary2.mean(),
                    'noi2': r1s2_drop.noi2.mean(),
                    'noe2': r1s2_drop.noe2.mean(),
                    'interest2': r1s2_drop.interest2.mean(),
                    'ctax2': r1s2_drop.ctax2.mean(),
                    'profit2': r1s2_drop.profit2.mean(),
                    'liquidAsset2': r1s2_drop.liquidAsset2.mean(),
                    'quickAsset2': r1s2_drop.quickAsset2.mean(),
                    'receivableS2': r1s2_drop.receivableS2.mean(),
                    'inventoryAsset2': r1s2_drop.inventoryAsset2.mean(),
                    'nonCAsset2': r1s2_drop.nonCAsset2.mean(),
                    'tanAsset2': r1s2_drop.tanAsset2.mean(),
                    'OnonCAsset2': r1s2_drop.OnonCAsset2.mean(),
                    'receivableL2': r1s2_drop.receivableL2.mean(),
                    'debt2': r1s2_drop.debt2.mean(),
                    'liquidLiabilities2': r1s2_drop.liquidLiabilities2.mean(),
                    'shortLoan2': r1s2_drop.shortLoan2.mean(),
                    'NCLiabilities2': r1s2_drop.NCLiabilities2.mean(),
                    'longLoan2': r1s2_drop.longLoan2.mean(),
                    'netAsset2': r1s2_drop.netAsset2.mean(),
                    'surplus2': r1s2_drop.surplus2.mean()}

In [785]:
test2 = r1s2_drop.replace(0, missing_fill_val)

In [787]:
# emplyee1,2

train2['employee1'] = train2['employee1'].replace(0, train2['employee1'].median())
train2['employee2'] = train2['employee2'].replace(0, train2['employee2'].median())

In [788]:
test2.employee1 = test2.employee1.astype('str').str.replace(",", "").astype('float')
test2.employee2 = test2.employee2.astype('str').str.replace(",", "").astype('float')

In [789]:
test2['employee1'] = test2['employee1'].replace(0, test2['employee1'].median())
test2['employee2'] = test2['employee2'].replace(0, test2['employee2'].median())

In [790]:
# ownerChange

train2['ownerChange'] = train2['ownerChange'].replace(0, 'empty')
test2['ownerChange'] = test2['ownerChange'].replace(0, 'empty')
ownerChange_encoder = LabelEncoder() # 인코더 생성

_ownerChange = list(train2['ownerChange'].values)
ownerChange_set = set(_ownerChange)
print(len(ownerChange_set))
ownerChange_encoder.fit(list(ownerChange_set)) # 인코딩

# train과 test에 숫자형 치환
train2['ownerChange'] = ownerChange_encoder.transform(train2['ownerChange'])
test2['ownerChange'] = ownerChange_encoder.transform(test2['ownerChange'])

3


In [745]:
# one-hot-encoding

In [793]:
train2 = pd.get_dummies(train2, columns=['sido'])
test2 = pd.get_dummies(test2, columns=['sido'])

In [794]:
train2 = pd.get_dummies(train2, columns=['instkind'])
test2 = pd.get_dummies(test2, columns=['instkind'])

In [795]:
train2 = pd.get_dummies(train2, columns=['ownerChange'])
test2 = pd.get_dummies(test2, columns=['ownerChange'])

In [796]:
train2.head(4)

inst_id  OC  sgg  bedCount      revenue1    salescost1          sga1  \
0        1   1   73     175.0  4.217530e+09  1.963479e+09  3.961135e+09   
1        3   1   32     410.0  1.270897e+10  1.963479e+09  1.021233e+10   
2        4   1   89     468.0  1.004522e+09  5.154837e+08  4.472197e+08   
3        7   1  141     353.0  7.250734e+10  1.963479e+09  7.067740e+10   

        salary1          noi1          noe1  ...  instkind_0  instkind_1  \
0  2.033835e+09  1.565244e+07  1.523624e+07  ...           0           0   
1  5.588052e+09  2.662286e+08  5.041395e+08  ...           0           0   
2  2.964023e+08  7.615600e+04  3.000000e+04  ...           0           0   
3  3.178605e+10  5.062231e+08  1.259568e+09  ...           0           0   

   instkind_3  instkind_4  instkind_5  instkind_6  instkind_7  ownerChange_0  \
0           0           0           1           0           0              0   
1           1           0           0           0           0              0   
2           0           0           1           0           0              0   
3           1           0           0           0           0              0   

   ownerChange_1  ownerChange_2  
0              0              1  
1              0              1  
2              0              1  
3              0              1  

[4 rows x 83 columns]

## 6. 데이터 분석(EDA)

### 히스토그램 확인 -> 정규화 

In [420]:
sns.pairplot(iris, diag_kind='hist')

plt.show()


NameError: name 'iris' is not defined

### 상관관계 분석후 파생변수 만들기 

In [ ]:
df = raw.corr()
# 그림 사이즈 지정
fig, ax = plt.subplots( figsize=(7,7) )

# 삼각형 마스크를 만든다(위 쪽 삼각형에 True, 아래 삼각형에 False)
mask = np.zeros_like(df, dtype=np.bool)
mask[np.triu_indices_from(mask)] = True

# 히트맵을 그린다
sns.heatmap(df, 
            cmap = 'RdYlBu_r', 
            annot = True,   # 실제 값을 표시한다
            mask=mask,      # 표시하지 않을 마스크 부분을 지정한다
            linewidths=.5,  # 경계면 실선으로 구분하기
            cbar_kws={"shrink": .5},# 컬러바 크기 절반으로 줄이기
            vmin = -1,vmax = 1   # 컬러바 범위 -1 ~ 1
           )  
plt.show()
[출처] 상관관계 분석 시각화 - correlation matrix (df.corr, sns.heatmap)

## 7. 모델 선택, 훈련, 검증

In [38]:
# 4개 모델 만들기!

# 1. 로지스틱 회귀 검증 ( 강사 )

In [ ]:
model=Sequential()
model.add(Dense(1, input_dim=1, activation='sigmoid'))
sgd=optimizers.SGD(lr=0.01)
model.compile(optimizer=sgd ,loss='binary_crossentropy',metrics=['accuracy'])
# 옵티마이저는 경사하강법 sgd를 사용합니다.
# 손실 함수(Loss function)는 binary_crossentropy(이진 크로스 엔트로피)를 사용합니다.
model.fit(X,y, batch_size=1, epochs=200, validation_split = 0.2)
# 주어진 X와 y데이터에 대해서 오차를 최소화하는 작업을 200번 시도합니다.

In [97]:
loss = hist.history['loss']
val_loss = hist.history['val_loss']
epochs = range(1, len(hist.history['accuracy']) + 1)

In [39]:
plt.plot( epochs, loss )
plt.plot( epochs, val_loss )
plt.title( 'Model loss' )
plt.xlabel('Epochs')
plt.ylabel('Loss')
plt.legend(['train', 'val'], loc = 'upper left')
plt.show()

NameError: name 'epochs' is not defined

In [99]:
print( '테스트 정확도 : {:.4f}'.format(model.evaluate(X_test, y_test )[1]))

30/1 [====================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================] - 0s 3ms/sample - loss: 0.1864 - accuracy: 1.0000
테스트 정확도 : 1.0000


In [ ]:
loss = hist.history['loss']
val_loss = hist.history['val_loss']
epochs = range(1, len(hist.history['accuracy']) + 1)

plt.plot( epochs, loss )
plt.plot( epochs, val_loss )
plt.title( 'Model loss' )
plt.xlabel('Epochs')
plt.ylabel('Loss')
plt.legend(['train', 'val'], loc = 'upper left')
plt.show()

print( '테스트 정확도 : {:.4f}'.format(model.evaluate(X_test, y_test )[1]))

# 2. 랜덤포레스트

현재 randomforest regressor의 최적의 parameter를 찾는 과정을 GridSearchCV를 활용하여 수행하려고 한다.

하지만, 전체 데이터를 사용하여서 최적의 paramter를 찾으려고 하는 경우, 시간이 매우 오래 걸리기 때문에

train data의 1%에 해당되는 데이터만 활용하여서 최적의 parameter 값을 찾고 이를 전체 train data에 학습시키려고 한다.

In [ ]:
X_train=train2[input_var]
random.seed(1217) #동일한 샘플링하기 위한 시드번호
train_list=random.sample(list(range(X_train.shape[0])), int(round(X_train.shape[0]*0.01,0)) )

X_train=train2[input_var]
X_train=X_train.iloc[train_list,:]
y_train=train2[target]
y_train=y_train.iloc[train_list,:]

X_test=test2[input_var]

X_train.shape, y_train.shape

In [ ]:
# Create the parameter grid based on the results of random search 
param_grid = {
    'max_features': [2,3,5],
    'min_samples_leaf': [2,3],
    'min_samples_split': [2,4,6],
    'n_estimators': [100, 200,500]
}
# Create a based model
rf = RandomForestRegressor(random_state=1217) # 랜덤포레스트 모델을 정의한다.
# Instantiate the grid search model
grid_search = GridSearchCV(estimator = rf, param_grid = param_grid) # GridSearchCV를 정의한다.

grid_search.fit(X_train, y_train)

grid_search.best_params_ #학습 이후 최적의 paramter를 출력

#해당 코드 실행시간 2분 ~ 3분 소요

In [ ]:
#전체 데이터로 적용
X_train=train2[input_var]
y_train=train2[target]

X_test=test2[input_var]

X_train.shape, y_train.shape, X_test.shape

In [ ]:
# best_params_를 통해서 나온 값을 투입
rf = RandomForestRegressor(max_features=3,min_samples_leaf=2,min_samples_split=2,n_estimators=500,random_state=1217)

rf.fit(X_train,y_train) #학습 

test['18~20_ride'] = rf.predict(X_test) #예측값 생성 후, test['18~20_ride']에 집어 넣는다.

test[['id','18~20_ride']].to_csv("dacon_base_middle2.csv",index=False) # id와 18~20_ride만 선택 후 csv 파일로 내보낸다

#해당 코드 소요 시간 5분

# 3. Xgboost

# 4. 딥러닝 사용하기!

In [ ]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

In [ ]:
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [ ]:
from keras import optimizers
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [ ]:
from keras import losses
from keras import metrics

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])

In [ ]:
x_val = x_train[:10000]
partial_x_train = x_train[10000:]
y_val = y_train[:10000]
partial_y_train = y_train[10000:]

In [ ]:
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['acc'])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

In [ ]:
>>> history_dict = history.history
>>> history_dict.keys()
[u'acc', u'loss', u'val_acc', u'val_loss']

In [ ]:
import matplotlib.pyplot as plt

history_dict = history.history
loss = history_dict['loss']
val_loss = history_dict['val_loss']

epochs = range(1, len(loss) + 1)

plt.plot(epochs, loss, 'bo', label='Training loss')  # ‘bo’는 파란색 점을 의미합니다.
plt.plot(epochs, val_loss, 'b', label='Validation loss') # ‘b’는 파란색 실선을 의미합니다.
plt.title('Training and validation loss')
plt.xlabel('Epochs')
plt.ylabel('Loss')
plt.legend()

plt.show()

In [ ]:
plt.clf() # 그래프를 초기화합니다.
acc = history_dict['acc']
val_acc = history_dict['val_acc']

plt.plot(epochs, acc, 'bo', label='Training acc')
plt.plot(epochs, val_acc, 'b', label='Validation acc')
plt.title('Training and validation accuracy')
plt.xlabel('Epochs')
plt.ylabel('Accuracy')
plt.legend()

plt.show()

In [ ]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)

In [ ]:
>>> model.predict(x_test)
array([[ 0.98006207]
       [ 0.99758697]
       [ 0.99975556]
       ...,
       [ 0.82167041]
       [ 0.02885115]
       [ 0.65371346]], dtype=float32)